src/arkode/arkode_arkstep.c

```diff
@@ -2099,12 +2187,216 @@ int arkStep_TakeStep_Z(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+     */
+    int nvec = 0;
+    for (int js = is + 1; js < step_mem->stages; ++js)
+    {
+      /* h sum_{j=i}^{s} A_{ji} \Lambda_{j} */
+      cvals[nvec] = -ark_mem->h * step_mem->Be->A[js][is];
+      Xvecs[nvec] = N_VGetSubvector_ManyVector(stage_values[js], 0);
+      nvec++;
+    }
+    cvals[nvec] = -ark_mem->h * step_mem->Be->b[is];
+    Xvecs[nvec] = lambda_np1;
+    nvec++;
+
+    /* h b_i \lambda_{n+1} + h sum_{j=i}^{s} A_{ji} \Lambda_{j} */
+    retval = N_VLinearCombination(nvec, cvals, Xvecs, Lambda_tmp);
+    if (retval != 0) { return (ARK_VECTOROP_ERR); }
+
+    /* Compute stage values \Lambda_i, \nu_i by applying f_{y,p}^T (which is what fe does in this case)  */
+    retval = step_mem->fe(ark_mem->tcur, sens_tmp, stage_values[is],
+                          ark_mem->user_data);
+    step_mem->nfe++;
+
+    /* The checkpoint was not found, so we need to recompute at least
+       this step forward in time. We first seek the last checkpointed step
+       solution, then recompute from there. */
+    if (retval > 0)
+    {
+      N_Vector checkpoint = N_VGetSubvector_ManyVector(ark_mem->tempv2, 0);
+      int64_t curr_step, start_step;
+      curr_step = start_step = ark_mem->adj_step_idx;
+
+      SUNErrCode errcode = SUN_ERR_CHECKPOINT_NOT_FOUND;
+      for (int64_t i = 0; i <= curr_step; ++i, --start_step)
+      {
+        SUNDIALS_MAYBE_UNUSED int64_t stop_step = curr_step + 1;
+        SUNLogDebug(ARK_LOGGER, "ARKODE::arkStep_TakeStep_ERK_Adjoint",
+                    "searching-for-checkpoint",
+                    "start_step = %li, stop_step = %li", start_step, stop_step);
+        sunrealtype checkpoint_t;
+        errcode =
+          SUNAdjointCheckpointScheme_LoadVector(ark_mem->checkpoint_scheme,
+                                                start_step, step_mem->stages,
+                                                /*peek=*/1, &checkpoint,
+                                                &checkpoint_t);
+        if (errcode == SUN_SUCCESS)
+        {
+          /* OK, now we have the last checkpoint that stored as (start_step, stages).
+             This represents the last step solution that was checkpointed. As such, we
+             want to recompute from start_step+1 to stop_step. */
+          start_step++;
+          sunrealtype t0 = checkpoint_t;
+          sunrealtype tf = ark_mem->tn;
+          SUNLogDebug(ARK_LOGGER, "ARKODE::arkStep_TakeStep_ERK_Adjoint",
+                      "start-recompute",
+                      "start_step = %li, stop_step = %li, t0 = %" SUN_FORMAT_G
+                      ", tf = %" SUN_FORMAT_G "",
+                      start_step, stop_step, t0, tf);
+          if (SUNAdjointStepper_RecomputeFwd(adj_stepper, start_step, t0, tf,
+                                             checkpoint))
+          {
+            return (ARK_ADJ_RECOMPUTE_FAIL);
+          }
+          SUNLogDebug(ARK_LOGGER, "ARKODE::arkStep_TakeStep_ERK_Adjoint",
```

Comment:
The function name is added automatically by the macro
```suggestion
          SUNLogDebug(ARK_LOGGER,
```